scene/resources/paragraph_layout.cpp

```diff
@@ -0,0 +1,501 @@
+
+	font = p_font;
+	dirty = true;
+}
+
+void ParagraphLayout::set_ot_features(const String &p_features) {
+
+	features.clear();
+
+	Vector<String> v_features = p_features.split(",");
+	for (int i = 0; i < v_features.size(); i++) {
+		hb_feature_t feature;
+		hb_feature_from_string(v_features[i].ascii().get_data(), -1, &feature);
+		features.push_back(feature);
+	}
+	dirty = true;
+}
+
+void ParagraphLayout::set_script(const String &p_script) {
+
+	script = p_script;
+	dirty = true;
+}
+
+void ParagraphLayout::set_lang(const String &p_lang) {
+
+	lang = p_lang;
+	dirty = true;
+}
+
+void ParagraphLayout::analyse_paragraph(bool p_force) {
+
+	if (!dirty && !p_force)
+		return;
+
+	lines.clear();
+
+	dirty = false;
+
+	if (max_width == -1) {
+		lines.push_back(_analyse_line(0, utf16_text.size()));
+	} else {
+		LineInfo line = _analyse_line(0, utf16_text.size());
+		if (line.visual_order.size() > 0) {
+			if (line.visual_order[line.visual_order.size() - 1]->get_offset() + line.visual_order[line.visual_order.size() - 1]->get_length() < max_width) {
+				lines.push_back(line);
+			} else {
+				int x_ofs = 0;
+				int x_last_safe_break = 0;
+				int last_safe_break = -1;
+				int line_start = 0;
+
+				for (int i = 0; i < line.logical_order.size(); i++) {
+					x_ofs += line.logical_order[i]->get_length();
+					if (x_ofs > max_width) {
+						if (last_safe_break - line_start > 0)
+							lines.push_back(_analyse_line(line_start, last_safe_break - line_start));
+						line_start = last_safe_break + 1;
+						x_ofs -= x_last_safe_break;
+					}
+					// TODO: update to Unicode® Standard Annex #14
+					if (u_isWhitespace(utf16_text[line.logical_order[i]->get_min_char_index()]) && (line.logical_order[i]->get_min_char_index() == line.logical_order[i]->get_max_char_index())) {
+						last_safe_break = line.logical_order[i]->get_min_char_index();
+						x_last_safe_break = x_ofs;
+					}
+				}
+				if (line_start < utf16_text.size()) {
+					lines.push_back(_analyse_line(line_start, utf16_text.size() - line_start));
+				}
+			}
+		}
+	}
+}
+
+LineInfo ParagraphLayout::_analyse_line(uint32_t p_start, uint32_t p_length) {
+
+	LineInfo line;
+	line.line_direction = DIR_MIXED;
+	UErrorCode err = U_ZERO_ERROR;
+
+	if ((utf16_text.size() == 0) || (p_start + p_length > utf16_text.size()) || (p_start < 0))
+		return line;
+
+	UBiDi *para = ubidi_openSized(p_length, 0, &err);
+	switch (base_direction) {
+		case DIR_LTR: {
+			ubidi_setPara(para, (const uint16_t *)&utf16_text[p_start], p_length, UBIDI_LTR, NULL, &err);
+		} break;
+		case DIR_RTL: {
+			ubidi_setPara(para, (const uint16_t *)&utf16_text[p_start], p_length, UBIDI_RTL, NULL, &err);
+		} break;
+		case DIR_AUTO: {
+			ubidi_setPara(para, (const uint16_t *)&utf16_text[p_start], p_length, UBIDI_DEFAULT_LTR, NULL, &err);
+		} break;
+	}
+	if (U_SUCCESS(err)) {
+		UBiDiDirection direction = ubidi_getDirection(para);
+		if (direction != UBIDI_MIXED) {
```

Comment:
I don’t think you gain much by handling non-mixed direction text specially since `ubidi_countRuns` already does pretty much the same thing, so it would be simpler to have one branch here.